### PR TITLE
Arweave uploading functionality

### PR DESCRIPTION
**Pull request by lexich**
Hi in https://github.com/metaplex/store/pull/45 we need to upload assets to Arweave and mint nft. As I understand there aren't ready actions for this purpose. I tried to migrate this functionality from https://github.com/metaplex-foundation/metaplex/blob/master/js/packages/web/src/actions/nft.tsx 
using `@metaplex/js` primitives.

### CODE REVIEW

File: src/actions/mintArveaweNFT.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+export function getAssetCostInfo(files: Map<string, Buffer>) {
+  const sizes = Array.from(files.values()).map((f) => f.byteLength);
+  return calculate(sizes);
+}
+
+export function createFilePack(metadata: MetadataJson, WebFile = File): File {
+  const filedata = {
+    animation_url: undefined,
+    attributes: metadata.attributes && metadata.attributes.length ? metadata.attributes : undefined,
```
Comment (kurpav) [inline src/actions/mintArveaweNFT.ts:45]:
```suggestion
    attributes: metadata.attributes?.length ? metadata.attributes : undefined,
```

File: src/actions/mintArveaweNFT.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+  const hex = hashSum.digest('hex');
+  return Buffer.from(hex);
+}
+
+export function getAssetCostInfo(files: Map<string, Buffer>) {
+  const sizes = Array.from(files.values()).map((f) => f.byteLength);
+  return calculate(sizes);
+}
+
+export function createFilePack(metadata: MetadataJson, WebFile = File): File {
```
Comment (kurpav) [inline src/actions/mintArveaweNFT.ts:42]: probably it makes sense to define interface or type for `WebFile` param

File: src/actions/mintArveaweNFT.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+  const [dataFile, dataMetadata] = await Promise.all([
+    file.arrayBuffer().then((d) => Buffer.from(d)),
+    fileMetadata.arrayBuffer().then((d) => Buffer.from(d)),
+  ]);
+  const files = new Map<string, Buffer>();
+  files.set(file.name, dataFile);
+  files.set(fileMetadata.name, dataMetadata);
+  return getAssetCostInfo(files);
+}
+export async function getCost(
```
Comment (kurpav) [inline src/actions/mintArveaweNFT.ts:81]: probably it makes sense to move cost calculation to a separate utils file

File: src/actions/mintArveaweNFT.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+export async function payForFiles(
+  {
+    file,
+    metadata,
+    wallet: w,
+  }: {
+    wallet: Wallet;
+    file: File;
+    metadata: MetadataJson;
+  },
```
Comment (kurpav) [inline src/actions/mintArveaweNFT.ts:101]: please define as type

File: src/actions/mintArveaweNFT.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  WebFile = File,
+) {
+  const fileMetadata = createFilePack(metadata, WebFile);
+  const [hashFile, hashMetadata] = await Promise.all([
+    convertFile2Hash(file),
+    convertFile2Hash(fileMetadata),
+  ]);
+  const files = new Map<string, File>();
+  files.set(file.name, file);
+  files.set(fileMetadata.name, fileMetadata);
```
Comment (kurpav) [inline src/actions/mintArveaweNFT.ts:111]: you have similar code in the line 72

File: src/actions/mintArveaweNFT.ts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    progress = () => {},
+  }: {
+    connection: Connection;
+    wallet: Wallet;
+    storage: ArweaveStorage;
+    file: File;
+    metadata: MetadataJson;
+    maxSupply: number;
+    progress?: (status: ENftProgress) => void;
+  },
```
Comment (kurpav) [inline src/actions/mintArveaweNFT.ts:153]: please define type

File: src/actions/mintArveaweNFT.ts [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+  signing_metadata_transaction,
+  sending_transaction_to_solana,
+  waiting_for_initial_confirmation,
+  waiting_for_final_confirmation,
+  uploading_to_arweave,
+  updating_metadata,
+  signing_token_transaction,
+}
+
+export async function convertFile2Hash(file: File): Promise<Buffer> {
```
Comment (m-sebastiyan) [inline src/actions/mintArveaweNFT.ts:29]: Is this intended to be exported? Same for other functions that are utils.

**Comment by kurpav** [comment]
I'd say it makes sense to add action to be able to pay for files and just combine it with other mint actions on the applications side. Since this action covers a specific case.

File: src/actions/mintArveaweNFT.ts [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+  {
+    connection,
+    wallet,
+    file,
+    metadata,
+    maxSupply,
+    storage,
+    progress = () => {},
+  }: {
+    connection: Connection;
```
Comment (m-sebastiyan) [inline src/actions/mintArveaweNFT.ts:146]: Would be great to use interfaces for params as in https://github.com/metaplex/js/blob/main/src/actions/initStore.ts

File: src/actions/mintArveaweNFT.ts [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+export async function getCost(
+  file: File,
+  metadata: MetadataJson,
+  fileMetadata = createFilePack(metadata),
+) {
+  const calcInfo = await getCostInfo(file, metadata, fileMetadata);
+  return calcInfo.solana * LAMPORTS_PER_SOL;
+}
+
+const ARWEAVE_WALLET = new PublicKey('6FKvsq4ydWFci6nGq9ckbjYMtnmaqAoatz5c9XWjiDuS');
```
Comment (m-sebastiyan) [inline src/actions/mintArveaweNFT.ts:90]: I think it's worth updating const in https://github.com/metaplex/js/blob/main/src/config.ts and utilizing it here.

File: src/actions/mintArveaweNFT.ts [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+  return Buffer.from(hex);
+}
+
+export function getAssetCostInfo(files: Map<string, Buffer>) {
+  const sizes = Array.from(files.values()).map((f) => f.byteLength);
+  return calculate(sizes);
+}
+
+export function createFilePack(metadata: MetadataJson, WebFile = File): File {
+  const filedata = {
```
Comment (m-sebastiyan) [inline src/actions/mintArveaweNFT.ts:43]:
```suggestion
 const { attributes, name, symbol, description, seller_fee_basis_points, image, external_url, properties: { files, category, creators } = metadata;
  const filedata = { attributes: attributes?.length ? metadata.attributes : undefined, name, symbol, description, seller_fee_basis_points, image, external_url, properties: { ... } }
```

File: src/actions/mintArveaweNFT.ts [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+import { Metadata, Creator, MetadataDataData, CreateMetadata } from '../programs/metadata';
+import { sendTransaction } from './transactions';
+import { updateMetadata } from './updateMetadata';
+import { createMasterEdition } from './createMasterEdition';
+
+export enum ENftProgress {
+  none,
+  minting,
+  preparing_assets,
+  signing_metadata_transaction,
```
Comment (m-sebastiyan) [inline src/actions/mintArveaweNFT.ts:20]: Is it possible to use camelCase?

**Review by m-sebastiyan — CHANGES_REQUESTED** [review]

File: src/actions/mintArveaweNFT.ts [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+        createAssociatedTokenAccountTx,
+        mintToTx,
+      ],
+    });
+
+    progress(ENftProgress.signing_metadata_transaction);
+
+    try {
+      await connection.confirmTransaction(txid, 'max');
+    } catch {}
```
Comment (b2kdaman) [inline src/actions/mintArveaweNFT.ts:212]: Missing error throw?

File: src/actions/mintArveaweNFT.ts [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+    progress(ENftProgress.waiting_for_initial_confirmation);
+
+    const arweaveResult = await storage.upload(
+      files as unknown as Map<string, Buffer>, // TODO: probably we need to update ArweaveStorage interface
+      mint.publicKey.toBase58(),
+      txid,
+    );
+    progress(ENftProgress.waiting_for_final_confirmation);
+
+    const metadataFile = arweaveResult.messages?.find((m) => m.filename === 'manifest.json');
```
Comment (b2kdaman) [inline src/actions/mintArveaweNFT.ts:227]: Move to const

File: src/actions/mintArveaweNFT.ts (hunk 6 above)
Comment (b2kdaman) [reply src/actions/mintArveaweNFT.ts:153]: Also use more readable name `setProgress`/`updateProgress`

File: src/actions/mintArveaweNFT.ts [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    const { mint, createMintTx, createAssociatedTokenAccountTx, mintToTx } =
+      await prepareTokenAccountAndMintTx(connection, wallet.publicKey);
+
+    const metadataPDA = await Metadata.getPDA(mint.publicKey);
+
+    const creators = metadata.properties.creators.map((c) => new Creator(c));
+    const metadataData = new MetadataDataData({
+      name: metadata.name,
+      symbol: metadata.symbol,
+      uri: ' '.repeat(64),
```
Comment (b2kdaman) [inline src/actions/mintArveaweNFT.ts:177]: Move to selfexplanatory const

File: src/actions/mintArveaweNFT.ts [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+      },
+      {
+        metadata: metadataPDA,
+        mint: mint.publicKey,
+        mintAuthority: wallet.publicKey,
+        updateAuthority: wallet.publicKey,
+        metadataData,
+      },
+    );
+    progress(ENftProgress.preparing_assets);
```
Comment (b2kdaman) [inline src/actions/mintArveaweNFT.ts:193]: Is there really a point to update progress on non async executions?

**Review by b2kdaman — CHANGES_REQUESTED** [review]

File: src/actions/mintArveaweNFT.ts (hunk 7 above)
Comment (b2kdaman) [reply src/actions/mintArveaweNFT.ts:29]: Better move these to utils

File: src/actions/mintArveaweNFT.ts (hunk 2 above)
Comment (lexich) [reply src/actions/mintArveaweNFT.ts:42]: Interface is define in `lib.dom.d.ts`

File: src/actions/mintArveaweNFT.ts (hunk 5 above)
Comment (lexich) [reply src/actions/mintArveaweNFT.ts:112]: Not exactly the same. On line 72 `Map<string, Buffer>` and here `Map<string, File>`

File: src/actions/mintArveaweNFT.ts (hunk 15 above)
Comment (lexich) [reply src/actions/mintArveaweNFT.ts:194]: Yes. This operation keeps a lot of time and we show this progress in UI

File: src/actions/mintArveaweNFT.ts (hunk 2 above)
Comment (lexich) [reply src/actions/mintArveaweNFT.ts]: It's defined in `lib/dom.d.ts` for nodejs env we can use `@web-std/file`

File: src/actions/mintArveaweNFT.ts (hunk 5 above)
Comment (lexich) [reply src/actions/mintArveaweNFT.ts]: Not exactly

File: src/actions/mintArveaweNFT.ts (hunk 7 above)
Comment (lexich) [reply src/actions/mintArveaweNFT.ts]: I've updated `getFileHash`` from `utils/crypto` for this purposes.

File: src/actions/mintArveaweNFT.ts (hunk 5 above)
Comment (kurpav) [reply src/actions/mintArveaweNFT.ts]: In this case you can use generic type

**Review by b2kdaman — APPROVED** [review]

**Comment by M4FFS** [comment]
Why did this get closed?

**Comment by kurpav** [comment]
@M4FFS The proposed changes cover a particular case and have parts that cover UI stuff. If a user needs more control over the mint process it makes sense to use low-level API on the application side.
So, we exposed the util function `prepareTokenAccountAndMintTx` in https://github.com/metaplex/js/pull/74

**Pull request closed without merging**